### PR TITLE
Demand Function updated

### PR DESCRIPTION
**Pull request by Gegas2231**
- Added time clashes, mandatory courses, program and semester constraints  …
- Changed the structure of the function so the iteration over all students occurs outside the function
- The test scenario is not working, optimization is infeasible, but the algorithm should work with proper data

### CODE REVIEW

**Comment by floswald** [comment]
looks good, thanks. wht you mean with *The test scenario is not working*? can we come up with an example that is runnable? would be good...

**Comment by Gegas2231** [comment]
With the test scenario I meant the example. It is working now, it still needs a little of luck to get a good draw, but with good data, the algorithm is working and all the constraints are good.

**Comment by floswald** [comment]
hi @Gegas2231 
i decided to merge julies version from #14 - now on master. can you please give a look at CourseMatch.jl/src/demand.jl and see if there is anything that needs to be ported over from your PR to that version or if that does all the things you want it to do already? thanks!

**Comment by Gegas2231** [comment]
The version is pretty much definitive. i will break down a couple of constraints to match the Students struct, but i have an exam tomorrow so i will probably work on this tomorrow afternoon.

**Comment by Gegas2231** [comment]
Hi Florian, I was completely absorbed by thesis so I could not make the quick fixes I said were left. The function now looks like what a final version of the function will be and matches a more complete version of the structure 'Student'. 

I do not know if this version should be merged into the master yet because first, the student.jl needs to be able to read all the information and properly categorize it, the current version of it is an early stage. However this is my final version of demand function and could be useful later if the project continues

File: src/demand.jl [hunk 1]
```diff
@@ -0,0 +1,171 @@
+"""
+    demand(price::Vector, Student::Student, p_neigh_parm::Int64=1)
```
Comment (floswald) [inline src/demand.jl]: we only need the first formulation, right? all info is contained in `Student`

File: src/demand.jl [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+Compute the total demand for classes based on the optimal individual bundles given preferences, budget and prices, and under constraints such as program, semester and time clashes using the `GurobiSolver()`.
+
+
+## Arguments
+
+Assume that N is the number of students, and C the number of classes that are available to choose from.
+
+- `price` : a column vector of dimension C x 1, which i-th element is the price that was assigned to class i.
+- `Student.pref` : an array containing S elements, and such that its n-th element is the (sparse) matrix representing student n preferences. Each of the matrices contained in that array should be a squared matrix of dimension C.
```
Comment (floswald) [inline src/demand.jl]: all of this needs to be moved to the doc string for `Student` in `src/student.jl`

File: src/demand.jl [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	dem = demand(price, Ind_pref, ind_budget, cap, time_const, mand_cour_const, prog_cour_const, sem_cour_const, elec_cour_const, num_elec_cour)
+
+	push!(mkt_demand, dem)
+	return mkt_demand
+end
+
+mkt_demand = Dict("ind_demands" => ans, "total_demand" => sum(ans) )
+
+"""
+
```
Comment (floswald) [inline src/demand.jl]: remove this blank line for the doc string to be associated to the function

**Comment by floswald** [comment]
manually merged and changed. latest version on branch `agg_dem`.

**Pull request closed without merging**